java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -1120,6 +1118,23 @@ private boolean hasLocallyPresentOptionalBinding(ResolvedBindings resolvedBindin
               && !getLocalExplicitBindings(keyFactory.unwrapOptional(resolvedBindings.key()).get())
                   .isEmpty();
         }
+
+        /**
+         * Returns {@code true} if {@code resolvedBindings} contains a binding for which there is
+         * an explicit present binding in this component that was resolved previously.
+         */
+        boolean hasLocalContributionBinding(ResolvedBindings resolvedBindings) {
+          return resolvedBindings
+                .contributionBindings()
+                .stream()
+                .anyMatch(contributionBinding ->
+                    !getLocalExplicitBindings(resolvedBindings.key()).isEmpty()
+                  && parentResolver.isPresent()
+                  && parentResolver.get().getPreviouslyResolvedBindings(
+                      resolvedBindings.bindingKey()).isPresent()
+                  && !parentResolver.get().getPreviouslyResolvedBindings(
```

Comment:
Why should this method care whether the parent bindings are empty?

The point is to determine whether we should bother resolving in this subcomponent despite the previously resolved bindings from the parent. We should if the subcomponent would resolve a binding potentially different than its parent, either validly (as in a multibinding with local contributions or a present optional binding where the parent is absent) or invalidly (as in a present optional binding where the parent's is also present or a different binding in a module installed only in the subcomponent).